### ISSUE TITLE
not well motivated by IDE support

### ISSUE THREAD

**Issue by michaelficarra** [bmeck/proposal-from-import#1]
The claimed motivation for this proposal is IDE support, but IDEs already have solutions for exactly this stated problem. @jridgewell has explained it very well  over on [TC39 Discourse](https://es.discourse.group/t/from-module-import-x/285/5).

![comment from Justin](https://user-images.githubusercontent.com/218840/78389365-7617bc80-7597-11ea-86eb-cf3fe2f138dc.png)

**Comment by bmeck** [comment]
That still causes left to right typing issues and collides with tab completion on RHS of that position which now is looking to be populated by module attributes. I agree things *can* be configured in various ways, but quick searches of social media do show programmers having problems with this. We shouldn't disregard all those who have issues, nor should we disregard the collision with future syntax possibilities. If a solution can be well propagated I think it would have done so by now. Even popular editors don't support this syntax in the way mentioned above out of the box.

**Comment by bmeck** [comment]
@michaelficarra to clarify, if people are consistently requesting this and tooling does not support this consistently/by default due to various reasons would gathering a listing reasons tools do not do the proposed workaround workflow be enough to provide evidence of motivation for you?

**Comment by michaelficarra** [comment]
If tools are opting to not do the workflow described by @jridgewell, hearing the reasons they don't could make me understand the motivation better, yes.

**Comment by bmeck** [comment]
@michaelficarra I've opened up https://github.com/bmeck/proposal-from-import/issues/2 to see if any tool does do that workflow and https://github.com/bmeck/proposal-from-import/issues/3 to gather comments about issues people are having. I will try and figure out how to directly sort responses from editors, but that is likely a difficult communication channel to establish.

**Comment by michaelficarra** [comment]
Apparently VS Code has the ability to define snippets in the way @jridgewell describes.

From @rauschma: https://twitter.com/rauschma/status/1221131422641070081

![tweet from rauschma](https://user-images.githubusercontent.com/218840/78400148-014e7d80-75ab-11ea-91f4-a0ce7f4cc069.png)

**Comment by bmeck** [comment]
@michaelficarra they autocomplete only after you populate the RHS, which is somewhat the issue at hand.

```mjs
// -> ltr typing
import {} from 'lib'|
// <- rewind to useful position
import {|} from 'lib'
// -> forward to useful position
import {} from 'lib'|
```

and

```mjs
// no useful completions
// there is a builtin snippet for vscode here but not by default completion
// snippet replays workflow and rewind/forward as above
import |
```

vs

```mjs
// -> continuous ltr typing
from 'lib' import |
```

**Comment by jridgewell** [comment]
(note, I don't currently use an IDE so I'm not getting auto-completion anyways…)

When using snippets, I've never been bothered by the backwards jumping with tab completion. I just hit tab and it goes directly to where I wanted.

**Comment by jasonwilliams** [comment]
There has been some discussion in the typescript team about using your package.json to see all the exports available, then filtering them down inside the `{}`, then auto completing the rest of the line based on what you pick.

https://github.com/microsoft/TypeScript/issues/31658

**Comment by andrewbranch** [comment]
The TL;DR of the TypeScript discussion @jasonwilliams just linked is that we would like to be able to offer useful completions for a whole import statement at `import |`, but doing so with reasonable performance can be difficult. We can quickly gather every possible export you might want to import, but complexities arising from symlinks (e.g. from lerna), node-style module resolution, and some TypeScript project-related things (project references and path mapping aliases) mean there may be many different possible module specifiers for each potential import, and discovering them all and choosing the best one can be expensive. (Normally, when we offer auto imports, we do all this work as lazily as possible, waiting until a specific import completion is focused within the list. That’s not an option here, because in this case the insertion is happening at the text insertion cursor, not somewhere else in the document.)

It’s still a feature we’re looking at, but it has proven more difficult than we initially thought it would be, and I’m not yet confident we can always provide a great UX here.

**Comment by rauschma** [comment]
Possible compromise – offer `fs` as a completion if it is used at least once somewhere in a code base (excluding `node_modules`):

```js
import * as fs from 'fs';
```

**Comment by bmeck** [comment]
@andrewbranch can we summarize that to the current syntax being problematic? It seems like this would be an issue with this proposal too if someone wrote `from import |`. and left the specifier out.

**Comment by andrewbranch** [comment]
I guess that would be problematic, but I think the idea is that users generally wouldn’t write that. I’m assuming the syntax order will have _some_ positive correlation with the order that most users incrementally write out an import statement, and it is much easier for us to give useful completions for exported identifiers when we already know the module specifier. I think in a hypothetical world where `from "..." import ...` exists, we would probably not give any completions at `from import |`, but I would be surprised if users found themselves in that situation frequently. After all, if a user wants to import something by a known name but doesn’t know the module specifier at all, they can use our existing auto-import functionality (just start typing an identifier reference and we’ll try to figure out the right import statement). If a user starts typing an import statement by hand, we think they likely have some idea how they’re going to finish it.

**Comment by bmeck** [comment]
@jridgewell ^

**Comment by hsablonniere** [comment]
Hey everyone,

I've been doing what @jridgewell describes for a few years now.
Because of this, I was thinking "we don't really need this in the language" as it can be solved/done in IDEs right now.

Following a discussion with @bmeck on Twitter, I realized that while this is possible with an *my* IDE and some config, most people struggle with this and most IDEs don't provide a default and simple way to do this. So I agree this proposal would benefit most users in a standard way.

In the mean time, and because I think a video demo is better than a description, here's what I have with WebStorm and a custom "live template" (it's how they call snippets).

https://user-images.githubusercontent.com/236342/118297615-b0fc5b80-b4de-11eb-9359-0dacda49ca56.mp4

## Details

* I trigger the live template with `ii` and then `tab` (I use `i` for simple imports).
* I have the same for `const {} = require();` with `rr`.
* I can validate a choice with tab or enter.
* I can go back to the import specifier with `shift+tab`.

I had to create this live template in the settings like this:

![Capture d’écran_2021-05-14_17-43-23](https://user-images.githubusercontent.com/236342/118297639-b8bc0000-b4de-11eb-8c2c-a17be0d0546d.png)

And I had to reorder the parts like this so it focuses in the right order:

![Capture d’écran_2021-05-14_17-43-35](https://user-images.githubusercontent.com/236342/118297662-bfe30e00-b4de-11eb-8777-4d17a4e95ffd.png)

Cheers :wink:

**Comment by andrewbranch** [comment]
As an update, TypeScript 4.3 is shipping with the feature mentioned earlier by @jasonwilliams.

https://devblogs.microsoft.com/typescript/announcing-typescript-4-3-rc/#import-statement-completions

<details>
<summary>Screen capture 📽</summary>

![GIF of import statement completions working in VS Code](https://user-images.githubusercontent.com/3277153/118300308-a3bb7e80-b496-11eb-8a30-927566115b93.gif)
</summary>
